### ISSUE TITLE
Validating Array objects for required fields, .exists() - doesn't seem to work

### ISSUE THREAD

**Issue by davida5** [ctavan/express-validator#531]
I am able to validate requests with latest version 5, works great, but facing one issue with arrays. Is there a way to force that a field must be present in array object? the exists() doesn't seem to work for arrays.
```
...
  return [
        check('items').exists(), //WORKS GREAT
        check('items.*._id').exists().isAlphanumeric(), //EXISTS() IS IGNORED  BUT WILL VALIDATE IF FIELD IS PRESENT :((
        check('items.*.item_type').exists().isIn(Item.ITEM_TYPES), //SAME HERE, EXISTS() IS IGNORED  BUT WILL VALIDATE IF FIELD IS PRESENT :((
check('items.*.update_status').exists().isInt().isIn(Constants.UPDATE_STATUS_VALUES).withMessage("must be one of:"+JSON.stringify(Constants.UPDATE_STATUS)),
        sanitize('items.*.long_desc').toString(),
    ];
```

**Comment by davida5** [comment]
Ok, I did some more digging, turns out exists() only checks that one of the items in the array has the field, instead of checking that each array item has the field. So I think it's a bug.
e.g. the below Body will pass the _id validation:
`{items:[{_id:"1"},{}]}`

**Comment by psychonetic** [comment]
@davida5 Did you write your own validator? I can confirm this behavior.

**Comment by davida5** [comment]
@psychonetic No I haven't used any custom validators. I'm glad you could confirm the behavior. I haven't found a workaround, I'm hoping this issue will resolve itself somehow in future versions...

**Comment by jonathonlui** [comment]
It's an issue with `check()`. Use one of specific checks `body()`, `params()`, etc. 

See discussion at https://github.com/ctavan/express-validator/issues/458 with the a partial fix implemented in https://github.com/ctavan/express-validator/pull/492.

**Comment by davida5** [comment]
Hey @jonathonlui thank you so much for taking the time and updating the thread. Could you also point out a simple way to do this body-only check, there doesn't seem to be any example on this.
I get an error: 

> checkBody(...).exists is not a function

with this code:
```
const {check, validationResult} = require('express-validator/check');
const checkBody = check(['body']);

ValidatorsHandler.prototype.sync = function () {
    return [
        checkBody('items').exists(), //must have at least one item
        checkBody('items.*._id').exists().isAlphanumeric(),
        checkBody('items.*.item_type').isIn(Item.ITEM_TYPES).withMessage("must be one of:"+JSON.stringify(Item.ITEM_TYPES)),
        checkBody('items.*.level').isInt(crit_level).withMessage(crit_level),
        checkBody('items.*.short_desc').isLength(crit_short_desc).withMessage(crit_short_desc),
        checkBody('items.*.title').isLength(crit_title).withMessage(crit_title),
        checkBody('items.*.update_status').isInt().isIn(Constants.UPDATE_STATUS_VALUES).withMessage("must be one of:"+JSON.stringify(Constants.UPDATE_STATUS)),
        sanitize('items.*.long_desc').toString(),
    ];
}
```

**Comment by gustavohenke** [comment]
Hey @davida5, the right name is `body`. See docs for more reference.

Also, while skimming through the issue, the feeling I have is that `*` is only working for the first results it encounters. Is it correct?

If so we'll need some serious bug fixing 🙂

**Comment by davida5** [comment]
@gustavohenke Yes it does check only the first item it encounters, a bit problematic as you can imagine 
:-)

So should I try the **_body_** fix or wait a bit more? I'm not sure what you mean by 

> the right name is body

could you point me into the place in the docs I wasn't able to find it.
Thanks for your help !

**Comment by jonathonlui** [comment]
@davida5 see https://github.com/ctavan/express-validator#check-api 

Simply change `check()` to `body()`. Example:


```javascript
...

exampleValidation = [
  body('items').exists().withMessage('missing'),
  body('items.*._id').isAlphanumeric(),
  body('items.*.item_type').exists().isIn([1, 2, 3]),
];

app.post('/example', exampleValidation , (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.mapped() });
  }
 ...
})

...
```

And it'll check each item. For example a request with 2 invalid item and 1 valid item results in 2 errors:

```http
POST /example HTTP/1.1
Accept: application/json, */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 83
Content-Type: application/json
Host: localhost:3000
User-Agent: HTTPie/0.9.9

{
    "items": [
        {
            "_id": "myid"
        },
        {
            "item_type": 1
        },
        {
            "_id": "gooditem",
            "item_type": 2
        }
    ]
}

HTTP/1.1 400 Bad Request
Connection: keep-alive
Content-Length: 184
Content-Type: application/json; charset=utf-8
Date: Mon, 05 Mar 2018 18:20:37 GMT
ETag: W/"b8-Uplw/sru0nh8L0Px5DDKkmqTICU"
X-Powered-By: Express

{
    "errors": {
        "items[0].item_type": {
            "location": "body",
            "msg": "Invalid value",
            "param": "items[0].item_type"
        },
        "items[1]._id": {
            "location": "body",
            "msg": "Invalid value",
            "param": "items[1]._id"
        }
    }
}
```

**Comment by midthun** [comment]
I have a similar issue using the `checkSchema` method. This is my schema: 
```
export const schema: ValidationSchema = {
  data: {
    in: ['body'],
    exists: true,
    errorMessage: 'Missing property \'data\' in \'body\''
  },
  'data.*.key': {
    in: ['body'],
    exists: true,
    errorMessage: 'Missing property \'key\' in array \'data\' in \'body\''
  },
  'data.*.value': {
    in: ['body'],
    exists: true,
    errorMessage: 'Missing property \'value\' in array \'data\' in \'body\''
  }
};
```

I am trying to make sure `data` exists on body and that each object in the `data` array have the `key` and `value` properties.

validating it like this: `checkSchema(schema)` with the body looking like: 
```
{
  data = []
}
```

passes validation, which I think it should not?

**Comment by gustavohenke** [comment]
Theoretically, it's correct @midthun -- if an item exists within `data`, then it must be valid according to those rules.
You could however add more constraints into `data`. Maybe `isLength` could help you

**Comment by midthun** [comment]
@gustavohenke I see, yes that makes sense. `isLength` on the `data` property helped! Thank you.

**Comment by gustavohenke** [comment]
Published [`v5.2.0`](https://github.com/express-validator/express-validator/releases/tag/v5.2.0) 🚢 with the fix
